jina/serve/runtimes/monitoring.py

```diff
@@ -14,9 +29,208 @@ def _setup_monitoring(self):
+            self._request_init_time[request.request_id] = time.time()
+        if self._receiving_request_histogram:
+            self._meter_request_init_time[request.request_id] = time.time()
+
+        if self._pending_requests_metrics:
+            self._pending_requests_metrics.inc()
+        if self._pending_requests_up_down_counter:
+            self._pending_requests_up_down_counter.add(
+                1, attributes=self._metric_labels
+            )
+
+    def _update_end_successful_requests_metrics(self, result: 'Request'):
+        if (
+                self._receiving_request_metrics
+        ):  # this one should only be observed when the metrics is succesful
+            init_time = self._request_init_time.pop(
+                result.request_id
+            )  # need to pop otherwise it stays in memory forever
+            self._receiving_request_metrics.observe(time.time() - init_time)
+        if (
+                self._receiving_request_histogram
+        ):  # this one should only be observed when the metrics is succesful
+            init_time = self._meter_request_init_time.pop(
+                result.request_id
+            )  # need to pop otherwise it stays in memory forever
+            self._receiving_request_histogram.record(
+                time.time() - init_time, attributes=self._metric_labels
+            )
+
+        if self._pending_requests_metrics:
+            self._pending_requests_metrics.dec()
+        if self._pending_requests_up_down_counter:
+            self._pending_requests_up_down_counter.add(
+                -1, attributes=self._metric_labels
+            )
+
+        if self._successful_requests_metrics:
+            self._successful_requests_metrics.inc()
+        if self._successful_requests_counter:
+            self._successful_requests_counter.add(1, attributes=self._metric_labels)
+
+        if self._sent_response_bytes:
+            self._sent_response_bytes.observe(result.nbytes)
+        if self._sent_response_bytes_histogram:
+            self._sent_response_bytes_histogram.record(
+                result.nbytes, attributes=self._metric_labels
+            )
+
+    def _update_end_failed_requests_metrics(self):
+        if self._pending_requests_metrics:
+            self._pending_requests_metrics.dec()
+        if self._pending_requests_up_down_counter:
+            self._pending_requests_up_down_counter.add(
+                -1, attributes=self._metric_labels
+            )
+
+        if self._failed_requests_metrics:
+            self._failed_requests_metrics.inc()
+        if self._failed_requests_counter:
+            self._failed_requests_counter.add(1, attributes=self._metric_labels)
+
+    def _update_end_request_metrics(self, result: 'Request'):
+        if result.status.code != jina_pb2.StatusProto.ERROR:
+            self._update_end_successful_requests_metrics(result)
+        else:
+            self._update_end_failed_requests_metrics()
```

Comment:
Should we also pop the Request entry in this method?

Reply:
I am just moving this method around